### DISCUSSION TITLE
Colorbar issue

### DISCUSSION THREAD

**Discussion by scanta2** [pyvista/pyvista#4092] [Q&A] [answered]
I don't have an example I can readily share, but I wanted to show a strange colorbar behavior.
depth1 are line segments, depths are integers in range(1,7), where each line segment's point is assigned a value in that range, the same value per line segment.

If I use the following code:
`plotter.add_mesh(depth1,scalars=depths,cmap=['green', 'blue', 'yellow', 'red', 'purple', 'cyan'], scalar_bar_args={'n_labels': 6, 'title': 'Depth'})`
I get this image:  (correct labels, wrong colors)
![screenshot (3)](https://user-images.githubusercontent.com/12899673/223220780-6486008d-82a9-4168-9cc1-7394e9c4cbac.png)

While
`plotter.add_mesh(depth1,scalars=depths,cmap=['green', 'blue', 'yellow', 'red', 'purple', 'cyan'])` shows the "wrong" labels, but correct colors:
![screenshot (4)](https://user-images.githubusercontent.com/12899673/223221064-0f654c5c-4d8e-4a60-bc68-35a509e56c2b.png)

**Comment by scanta2** [comment, marked as answer]
I have an example:
```py
import numpy as np
import random
import pyvista as pv
from itertools import chain
plotter = pv.Plotter()
plotter.show()
def segment():
    return tuple(np.array([random.uniform(0,1) for i in range(3)]) for i in range(2))
colors = ['green', 'blue', 'yellow', 'red', 'purple', 'cyan']
points = [segment() for i in range(100)]
depths = [[il%6 for i in range(len(l))] for il, l in enumerate(points)]
depths = list(chain.from_iterable(depths))
lines = [[len(l), 2*il, 2*il+1] for il, l in enumerate(points)] 
lines = [len(lines), *(chain.from_iterable(lines))]
points = list(chain.from_iterable(points))
polydata = pv.PolyData(points, lines=lines)
actor = plotter.add_mesh(polydata, scalars=depths, cmap=colors, scalar_bar_args={'n_labels': 6, 'title': 'Depth'})
```

**Reply by adeak** [reply]
Can you try passing `categories=True` to `add_mesh()`? I think this is exactly what it's supposed to handle.

**Reply by scanta2** [reply]
Thanks for the suggestion. It made no difference, though.

**Reply by adeak** [reply]
Indeed, I checked in the meantime, sorry. I can get the right colours if I do `scalars=np.array(colors)[depths]` instead, i.e. directly pass an array of colours for each point. But then the colorbar annotations are weird, they are the string names of the colours but their layout is (to me inexplicably) weird.

**Reply by banesullivan** [reply]
It appears to be an issue with `n_colors`. I noticed that passing any `scalar_bar_args` causes the issue. Adding `n_colors=256` to the  `scalar_bar_args ` seems to resolve this.

I think it's due to the default not being set if passing `scalar_bar_args`:

https://github.com/pyvista/pyvista/blob/50c5eda822b2c564ce18e0745b8126ec4092875d/pyvista/plotting/mapper.py#L608-L609

```py
import numpy as np
import random
import pyvista as pv
from itertools import chain

def segment():
    return tuple(np.array([random.uniform(0,1) for i in range(3)]) for i in range(2))
colors = ['green', 'blue', 'yellow', 'red', 'purple', 'cyan']
points = [segment() for i in range(100)]
depths = [[il%6 for i in range(len(l))] for il, l in enumerate(points)]
depths = list(chain.from_iterable(depths))
lines = [[len(l), 2*il, 2*il+1] for il, l in enumerate(points)] 
lines = [len(lines), *(chain.from_iterable(lines))]
points = list(chain.from_iterable(points))
polydata = pv.PolyData(points, lines=lines)


plotter = pv.Plotter()
actor = plotter.add_mesh(polydata, scalars=depths, cmap=colors, scalar_bar_args={'n_labels': 6, 'n_colors': 256, 'title': 'Depth'})
plotter.show()
```

![screenshot (10)](https://user-images.githubusercontent.com/22067021/225423386-1099b39d-ce92-43fa-931a-108ad1f4c5cf.png)

**Reply by scanta2** [reply]
> It appears to be an issue with `n_colors`. I noticed that passing any `scalar_bar_args` causes the issue. Adding `n_colors=256` to the `scalar_bar_args ` seems to resolve this.
> 
> I think it's due to the default not being set if passing `scalar_bar_args`:
> 
> https://github.com/pyvista/pyvista/blob/50c5eda822b2c564ce18e0745b8126ec4092875d/pyvista/plotting/mapper.py#L608-L609
> 
> ```python
> import numpy as np
> import random
> import pyvista as pv
> from itertools import chain
> 
> def segment():
>     return tuple(np.array([random.uniform(0,1) for i in range(3)]) for i in range(2))
> colors = ['green', 'blue', 'yellow', 'red', 'purple', 'cyan']
> points = [segment() for i in range(100)]
> depths = [[il%6 for i in range(len(l))] for il, l in enumerate(points)]
> depths = list(chain.from_iterable(depths))
> lines = [[len(l), 2*il, 2*il+1] for il, l in enumerate(points)] 
> lines = [len(lines), *(chain.from_iterable(lines))]
> points = list(chain.from_iterable(points))
> polydata = pv.PolyData(points, lines=lines)
> 
> 
> plotter = pv.Plotter()
> actor = plotter.add_mesh(polydata, scalars=depths, cmap=colors, scalar_bar_args={'n_labels': 6, 'n_colors': 256, 'title': 'Depth'})
> plotter.show()
> ```
> 
> ![screenshot (10)](https://user-images.githubusercontent.com/22067021/225423386-1099b39d-ce92-43fa-931a-108ad1f4c5cf.png)

Thanks, @banesullivan , but it does not look correct. The numbers do not match the color breakdowns, 1 should be green and rendered somewhere above the green rectangle, 2 should be blue, etc up to 6 cyan

**Reply by scanta2** [reply]
Actually, you are right, it's 0 to 5 because of how I used il%6, but the labels are still misaligned.

**Reply by banesullivan** [reply]
Yep, for these categorical colormaps I often add a -0.5 and +0.5 padding to the ends of the `clim` to have the labels center on the colors

**Reply by scanta2** [reply]
This is what happens if I add `clim = [0.5, 6.5]` to my code. How do I force the color bar to render integers (or even floats) from 1 to 6 instead?
![screenshot](https://user-images.githubusercontent.com/12899673/225431660-c1608375-afb1-4c72-b142-6681d5e3a5c9.png)

**Reply by banesullivan** [reply]
Shoot, that's right. You may need to use annotations for the labels. See these examples:

- https://docs.pyvista.org/api/plotting/_autosummary/pyvista.LookupTable.annotations.html#annotations
- https://docs.pyvista.org/examples/02-plot/scalar-bars.html

**Reply by scanta2** [reply]
@banesullivan Great, this worked.

```
annotations = {i: str(i) for i in range(1,7)}
raysActor = plotter.add_mesh(depth1,scalars=depths, annotations=annotations, clim=[0.5, 6.5], cmap=['green', 'blue', 'yellow', 'red', 'purple', 'cyan'], categories=True, scalar_bar_args={'n_colors': 256, 'n_labels': 0, 'title': 'Depth'})
```


I used 6 annotations, and no labels, and it did the trick.

Do you want this issue still open to take care of the `n_colors` missing default, or should I close it? Thanks for the help.

**Reply by banesullivan** [reply]
I've marked the discussion as solved -- but we do need to fix the `n_colors` issue in the code. I think this warrants a separate issue
